### ISSUE TITLE
NPE when auto.create and incorrect access privileges

### ISSUE THREAD

**Issue by lancezaklan** [confluentinc/kafka-connect-jdbc#646]
When the role used for accessing the database did not have the correct privileges to create a table, during the attempted creation of the table the jdbc connector threw an NPE. While the privileges where definitely not correct, crashing the sink seems incorrect.

As I read the code in DBStructure:createOrAmendIfNecessary, there is an error in the processing of a failed table create. Specifically either the cache update (tableDefns.refresh()) should probably defend against the case in which the tableDefns cache can not be updated because there is no table in the data base, or if the create table fails with sql exception, there should be no further cache manipulations, rather the original exception should log the exception (as an error) and re-throw.
```
    if (tableDefns.get(connection, tableId) == null) {
      // Table does not yet exist, so attempt to create it ...
      try {
        create(config, connection, tableId, fieldsMetadata);
      } catch (SQLException sqle) {
        log.warn("Create failed, will attempt amend if table already exists", sqle);
        try {
          TableDefinition newDefn = tableDefns.refresh(connection, tableId);
          if (newDefn == null) {
            throw sqle;
          }
        } catch (SQLException e) {
          throw sqle;
        }
      }
    }
    return amendIfNecessary(config, connection, tableId, fieldsMetadata, config.maxRetries);
```
probably should look like this (as the database does not have a table corresponding to the tableId and doesn't need to be made coherent, it's already in the right state when the create fails).
``` 
   if (tableDefns.get(connection, tableId) == null) {
      // Table does not yet exist, so attempt to create it ...
      try {
        create(config, connection, tableId, fieldsMetadata);
      } catch (SQLException sqle) {
        log.error("Create failed " + sqle.message(), sqle);
          throw sqle;
      }
    return amendIfNecessary(config, connection, tableId, fieldsMetadata, config.maxRetries);
 ```
if the SQLException wasn't used as part of the retry mechanism in the write method, then a try catch wouldn't even be necessary.  As it is, because it is very useful to get the exact message in a log error message, I suggest changing the code to put out an error log right at the original point of failure.

Here's the stack trace from the crash which doesn't directly show the failed privilege because of the nested try/catch. It does show the crash because the database did not have the table that had failed to create. The NPE was the result of dielect.describeTable(connection, tableId) returning null (which is proper as there is no table to describe).
```
org.apache.kafka.connect.errors.ConnectException: Exiting WorkerSinkTask due to unrecoverable exception.
    at org.apache.kafka.connect.runtime.WorkerSinkTask.deliverMessages(WorkerSinkTask.java:560)
    at org.apache.kafka.connect.runtime.WorkerSinkTask.poll(WorkerSinkTask.java:321)
    at org.apache.kafka.connect.runtime.WorkerSinkTask.iteration(WorkerSinkTask.java:224)
    at org.apache.kafka.connect.runtime.WorkerSinkTask.execute(WorkerSinkTask.java:192)
    at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:175)
    at org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:219)
    at java.util.concurrent.Executors.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
    at java.util.concurrent.ThreadPoolExecutor.run(ThreadPoolExecutor.java:624)
    at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NullPointerException
    at com.nike.eapp.kafka.connector.jdbc.util.TableDefinitions.refresh(TableDefinitions.java:85)
    at com.nike.eapp.kafka.connector.jdbc.sink.DbStructure.createOrAmendIfNecessary(DbStructure.java:64)
    at com.nike.eapp.kafka.connector.jdbc.sink.BufferedRecords.add(BufferedRecords.java:84)
    at com.nike.eapp.kafka.connector.jdbc.sink.JdbcDbWriter.write(JdbcDbWriter.java:65)
    at com.nike.eapp.kafka.connector.jdbc.sink.JdbcSinkTask.put(JdbcSinkTask.java:73)
    at org.apache.kafka.connect.runtime.WorkerSinkTask.deliverMessages(WorkerSinkTask.java:538)
    ... 10 more
 ```

**Comment by rhauch** [comment]
Thanks for logging the issue, @lancezaklan. What version of the JDBC sink connector were you using? It looks like 5.1.0 based upon the line numbers, but it'd be great to know for sure.

The NPE is definitely a problem and we should fix it, but I think the better fix for the NPE is for `TableDefinitions.refresh(...)` to check whether the `dbTable` is null, in which case the method should simply return null rather than attempting to dereference the pointer. The JavaDoc even states that this case should be handled, but it's clear from the code that it is not properly handled, and thus the NPE. 

Interestingly, if we were to make that fix, then the current `DbStructures.createOrAmendIfNecessary(...)` logic would behave correctly in the scenario you describebe by rethrowing the *original exception* from the attempt to create the table (e.g., no permissions).

I don't believe the `DbStructures.createOrAmendIfNecessary(...)` code should be changed, however. 
The `DbStructures.createOrAmendIfNecessary(...)` method is written this way to handle a possible but unlikely race condition where the table doesn't yet exist when the connector looks in the cache (e.g., `tableDefns.get(connection, tableId)` returns false on the first line of the fragment you included above), but is created by another DB client just before the code attempts to create the table. In such a case, the "CREATE TABLE" statement would result in a SQLException (e.g., "table already exists"), and the current code refreshes its cache to see if the table does or does not exist. If it does exist, the code goes on to attempt to amend the table; if it does not yet exist, then the current code throws the *original* exception from the "CREATE TABLE" attempt -- which in the scenario you describe is the permissions error.

Your suggested change to the `DbStructures.createOrAmendIfNecessary(...)` method is simpler, but it would not properly handle the unlikely but possible race condition I mentioned.

Do you have any interest in creating a PR (against the `5.0.x` branch) to handle the null `dbTable` reference in the `TableDefinitions.refresh(...)` method?

**Comment by kkonstantine** [comment]
Actual PR targeting this issue is: https://github.com/confluentinc/kafka-connect-jdbc/pull/649

**Comment by gharris1727** [comment]
This was fixed in https://github.com/confluentinc/kafka-connect-jdbc/pull/706
The fix should be avaliable in CP 5.0.4, 5.1.4, 5.2.4, 5.3.2, 5.4.0
